src/backend/distributed/planner/deparse_shard_query.c

```diff
@@ -350,6 +327,93 @@ UpdateRelationsToLocalShardTables(Node *node, List *relationShardList)
 }
 
 
+/*
+ * ReplaceRelationConstraintByShardConstraint replaces given OnConflictExpr's
+ * constraint id with constraint id of the corresponding shard.
+ */
+static bool
+ReplaceRelationConstraintByShardConstraint(List *relationShardList,
+										   OnConflictExpr *onConflict)
+{
+	Oid constraintId = onConflict->constraint;
+
+	if (!OidIsValid(constraintId))
+	{
+		return false;
+	}
+
+	Oid constraintRelationId = InvalidOid;
+
+	HeapTuple heapTuple = SearchSysCache1(CONSTROID, ObjectIdGetDatum(constraintId));
+	if (HeapTupleIsValid(heapTuple))
+	{
+		Form_pg_constraint contup = (Form_pg_constraint) GETSTRUCT(heapTuple);
+
+		constraintRelationId = contup->conrelid;
+		ReleaseSysCache(heapTuple);
+	}
+
+	/*
+	 * We can return here without calling the walker function, since we know there
+	 * will be no possible tables or constraints after this point, by the syntax.
+	 */
+	if (!OidIsValid(constraintRelationId))
+	{
+		ereport(ERROR, (errmsg("Invalid relation id (%u) for constraint: %s",
+							   constraintRelationId, get_constraint_name(constraintId))));
+		return false;
```

Comment:
We don't need to have `return` after `ereport(ERROR`, as the transaction is terminated and rollbacked immediately.

It  is also a little confusing, so let's remove